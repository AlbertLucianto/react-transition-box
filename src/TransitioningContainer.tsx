import React, {
  useMemo,
  useRef,
  useState,
} from 'react';
import {
  createContainerStyle,
  innerStyle,
  objectStyle,
} from './styles';
import useResizeListener from './useResizeListener';

interface ITransitioningContainerProps {
  children: JSX.Element|string;
  duration: number;
  [props: string]: any;
}

function TransitioningContainer({
  children,
  duration,
  style: propStyle,
  ...restProps
}: ITransitioningContainerProps) {
  const ref = useRef<HTMLObjectElement>(null);
  const [style, setStyle] = useState({});
  const containerStyle = useMemo(() => createContainerStyle(duration), [duration]);

  const mergedStyle = useMemo(() => ({
    ...style,
    ...containerStyle,
    ...(propStyle || {}),
  }), [style, propStyle]);

  useResizeListener(ref, setStyle);

  return (
    <div style={mergedStyle} {...restProps}>
      <div style={innerStyle}>
        <object
          ref={ref}
          data="about:blank"
          aria-hidden={true}
          aria-label="resize-listener"
          style={objectStyle}
          tabIndex={-1}
          type="text/html"
        />
        {children}
      </div>
    </div>
  );
}

export default TransitioningContainer;
