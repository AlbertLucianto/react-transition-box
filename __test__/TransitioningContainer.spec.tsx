import React, { useMemo } from 'react';
import { cleanup, render } from 'react-testing-library';
import TransitioningContainer from '../src';

afterEach(cleanup);

const waitFor = async (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});

interface IFooProps {
  width: number;
  height: number;
}

function Foo({ width, height }: IFooProps) {
  const style = useMemo(() => ({ width, height }), [width, height]);
  return (
    <TransitioningContainer
      duration={100}
      className="foo__container"
    >
      <div style={style} />
    </TransitioningContainer>
  );
}

describe('TransitioningContainer', () => {
  test('should render correctly', async (done) => {
    const container = render(<Foo
      width={100}
      height={200}
    />);

    const el: HTMLDivElement = container.baseElement.querySelector('.foo__container');
    expect(el.style.width).toBe('0px');
    expect(el.style.height).toBe('0px');

    await waitFor(100);
    expect(el.style.width).toBe(100);
    expect(el.style.height).toBe(200);

    container.rerender(<Foo
      width={150}
      height={100}
    />);

    expect(el.style.width).toBe(100);
    expect(el.style.height).toBe(200);

    await waitFor(100);
    expect(el.style.width).toBe(150);
    expect(el.style.height).toBe(100);

    done();
  });
});
