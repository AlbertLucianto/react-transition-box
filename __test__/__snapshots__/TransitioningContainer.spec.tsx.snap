// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitioningContainer should render correctly 1`] = `
<div
  className="foo__container"
  style={
    Object {
      "transition": "
    width 100ms ease,
    height 100ms ease
  ",
    }
  }
  timingFunction="ease"
>
  <div
    style={
      Object {
        "height": "fit-content",
        "position": "relative",
        "width": "fit-content",
      }
    }
  >
    <object
      aria-hidden={true}
      aria-label="resize-listener"
      data="about:blank"
      style={
        Object {
          "display": "block",
          "height": "100%",
          "left": 0,
          "opacity": 0,
          "overflow": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "width": "100%",
          "zIndex": -1,
        }
      }
      tabIndex={-1}
      type="text/html"
    />
    <div
      className="abc"
      style={
        Object {
          "height": 200,
          "width": 100,
        }
      }
    />
  </div>
</div>
`;

exports[`TransitioningContainer should render correctly 2`] = `
<div
  className="foo__container"
  style={
    Object {
      "height": 200,
      "transition": "
    width 100ms ease,
    height 100ms ease
  ",
      "width": 100,
    }
  }
  timingFunction="ease"
>
  <div
    style={
      Object {
        "height": "fit-content",
        "position": "relative",
        "width": "fit-content",
      }
    }
  >
    <object
      aria-hidden={true}
      aria-label="resize-listener"
      data="about:blank"
      style={
        Object {
          "display": "block",
          "height": "100%",
          "left": 0,
          "opacity": 0,
          "overflow": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "width": "100%",
          "zIndex": -1,
        }
      }
      tabIndex={-1}
      type="text/html"
    />
    <div
      className="abc"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;

exports[`TransitioningContainer should render correctly 3`] = `
<div
  className="foo__container"
  style={
    Object {
      "height": 200,
      "transition": "
    width 100ms ease,
    height 100ms ease
  ",
      "width": 100,
    }
  }
  timingFunction="ease"
>
  <div
    style={
      Object {
        "height": "fit-content",
        "position": "relative",
        "width": "fit-content",
      }
    }
  >
    <object
      aria-hidden={true}
      aria-label="resize-listener"
      data="about:blank"
      style={
        Object {
          "display": "block",
          "height": "100%",
          "left": 0,
          "opacity": 0,
          "overflow": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "width": "100%",
          "zIndex": -1,
        }
      }
      tabIndex={-1}
      type="text/html"
    />
    <div
      className="abc"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;
